JSC::Yarr::Flags in RegularExpression

annevk opened the pull request:
#### 1083694fd4b7f1755a24f1b99b4eb01acbeec888
<pre>
JSC::Yarr::Flags in RegularExpression
<a href="https://bugs.webkit.org/show_bug.cgi?id=255982">https://bugs.webkit.org/show_bug.cgi?id=255982</a>
rdar://108836530

Reviewed by Ross Kirsling.

Replace the MultilineMode, TextCaseSensitivity, and UnicodeMode enums
with JSC::Yarr:Flags and add an assert that prevents using the other
flags mainly to ensure folks use them consciously.

Also remove the JSC::Yarr prefix when it&apos;s not needed.

* Source/JavaScriptCore/inspector/ContentSearchUtilities.cpp:
(Inspector::ContentSearchUtilities::createRegularExpressionForSearchString):
* Source/JavaScriptCore/yarr/RegularExpression.cpp:
(JSC::Yarr::RegularExpression::Private::compile):
* Source/JavaScriptCore/yarr/RegularExpression.h:
* Source/WebCore/html/BaseTextInputType.cpp:
(WebCore::BaseTextInputType::patternMismatch const):
* Source/WebCore/html/EmailInputType.cpp:
(WebCore::isValidEmailAddress):
* Source/WebCore/inspector/InspectorStyleSheet.cpp:
(WebCore::selectorsFromSource):
* Source/WebCore/page/LocalFrame.cpp:
(WebCore::createRegExpForLabels):
* Source/WebCore/platform/graphics/avfoundation/CDMPrivateMediaSourceAVFObjC.mm:
(WebCore::CDMPrivateMediaSourceAVFObjC::parseKeySystem):
* Source/WebCore/platform/mac/StringUtilities.mm:
(WebCore::stringMatchesWildcardString):
* Source/WebKitLegacy/mac/WebView/WebHTMLRepresentation.mm:
(regExpForLabels):

Canonical link: <a href="https://commits.webkit.org/264136@main">https://commits.webkit.org/264136@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b5302cff72ae421e08ea444f38572cd228d62d4f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/6768 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/6986 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/7170 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/8357 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/7021 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/6766 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/8265 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/6929 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/9920 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/6885 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/7579 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/6210 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/8452 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/4987 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/6129 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/13939 "1 flakes 1 missing results 425 failures") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/5709 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/6703 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/6209 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/8982 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/6346 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/6697 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/5488 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/6905 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/6082 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/1612 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1616 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/10260 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/7095 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/6457 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/1746 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

rkirsling commented on Source/JavaScriptCore/yarr/RegularExpression.h:
```diff
@@ -29,14 +29,14 @@
-enum TextCaseSensitivity { TextCaseSensitive, TextCaseInsensitive };
-enum UnicodeMode { UnicodeUnawareMode, UnicodeAwareMode };
+enum class MultilineMode : bool { Disabled, Enabled };
+enum class TextCaseSensitivity : bool { Sensitive, Insensitive };
```
It would be neat if there were a way to avoid repeating (forms of) the word "sensitive"... that said, I suppose `CaseSensitive::{Yes,No}` or `IgnoreCase::{No,Yes}` might make the fact that "sensitive" is the default more confusing. Maybe the word `Text` could be dropped, at least?

rkirsling reviewed (approved):
r=me, a nice improvement.

Regarding OptionSet: The compile error is probably just wanting you to explicitly write `OptionSet<Flags>` before the braces. I think the "could be more elegant" part is fine.

rkirsling commented on Source/JavaScriptCore/yarr/RegularExpression.cpp:
```diff
@@ -38,29 +37,25 @@ namespace JSC { namespace Yarr {
 class RegularExpression::Private : public RefCounted<RegularExpression::Private> {
 public:
-    static Ref<Private> create(StringView pattern, TextCaseSensitivity caseSensitivity, MultilineMode multilineMode, UnicodeMode unicodeMode)
+    static Ref<Private> create(StringView pattern, OptionSet<JSC::Yarr::Flags> flags)
```
You shouldn't need to use the Yarr prefix anywhere here since you're already in that namespace -- I'm surprised there's so much existing code doing so.

rkirsling commented on Source/JavaScriptCore/yarr/RegularExpression.cpp:
```diff
@@ -38,29 +37,25 @@ namespace JSC { namespace Yarr {
+        ASSERT(!flags.contains(JSC::Yarr::Flags::Global));
+        ASSERT(!flags.contains(JSC::Yarr::Flags::DotAll));
+        ASSERT(!flags.contains(JSC::Yarr::Flags::UnicodeSets));
+        ASSERT(!flags.contains(JSC::Yarr::Flags::Sticky));
```
I think you could write this like
```cpp
ASSERT(!(flags - { Flags::IgnoreCase, Flags::Multiline, Flags::Unicode }));
```
and that way it'd remain valid even if more flags are added.

rkirsling commented on Source/JavaScriptCore/yarr/RegularExpression.h:
```diff
@@ -25,18 +25,16 @@
     WTF_MAKE_FAST_ALLOCATED;
 public:
-    explicit RegularExpression(StringView, TextCaseSensitivity = TextCaseSensitive, MultilineMode = MultilineDisabled, UnicodeMode = UnicodeUnawareMode);
+    explicit RegularExpression(StringView, OptionSet<JSC::Yarr::Flags> = { });
```
Ditto on the namespace comment.

rkirsling commented on Source/JavaScriptCore/inspector/ContentSearchUtilities.cpp:
```diff
@@ -143,7 +143,7 @@ RegularExpression createRegularExpressionForSearchString(const String& searchStr
         break;
     }
-    return RegularExpression(pattern, caseSensitive ? TextCaseSensitive : TextCaseInsensitive);
+    return caseSensitive ? RegularExpression(pattern) : RegularExpression(pattern, { JSC::Yarr::Flags::IgnoreCase });
```
This code isn't in the Yarr namespace but there is a `using namespace JSC::Yarr;` at the top of the file.

annevk replied:
I'll fix this to be locally consistent, but this is against the coding guidelines, right?

rkirsling replied:
Yeah...it's not something I'd recommend newly adding to files.

rkirsling reviewed (approved):
r=me with comments

annevk committed:
JSC::Yarr::Flags in RegularExpression
https://bugs.webkit.org/show_bug.cgi?id=255982
rdar://108836530

Reviewed by Ross Kirsling.

Replace the MultilineMode, TextCaseSensitivity, and UnicodeMode enums
with JSC::Yarr:Flags and add an assert that prevents using the other
flags mainly to ensure folks use them consciously.

Also remove the JSC::Yarr prefix when it's not needed.

* Source/JavaScriptCore/inspector/ContentSearchUtilities.cpp:
(Inspector::ContentSearchUtilities::createRegularExpressionForSearchString):
* Source/JavaScriptCore/yarr/RegularExpression.cpp:
(JSC::Yarr::RegularExpression::Private::compile):
* Source/JavaScriptCore/yarr/RegularExpression.h:
* Source/WebCore/html/BaseTextInputType.cpp:
(WebCore::BaseTextInputType::patternMismatch const):
* Source/WebCore/html/EmailInputType.cpp:
(WebCore::isValidEmailAddress):
* Source/WebCore/inspector/InspectorStyleSheet.cpp:
(WebCore::selectorsFromSource):
* Source/WebCore/page/LocalFrame.cpp:
(WebCore::createRegExpForLabels):
* Source/WebCore/platform/graphics/avfoundation/CDMPrivateMediaSourceAVFObjC.mm:
(WebCore::CDMPrivateMediaSourceAVFObjC::parseKeySystem):
* Source/WebCore/platform/mac/StringUtilities.mm:
(WebCore::stringMatchesWildcardString):
* Source/WebKitLegacy/mac/WebView/WebHTMLRepresentation.mm:
(regExpForLabels):

Canonical link: https://commits.webkit.org/264136@main
